crates/matrix-sdk/src/room/timeline/mod.rs

```diff
@@ -0,0 +1,217 @@
+//! A high-level view into a room's contents.
+//!
+//! See [`Timeline`] for details.
+
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use futures_signals::signal_vec::{MutableVec, SignalVec};
+use matrix_sdk_base::deserialized_responses::EncryptionInfo;
+use ruma::{
+    assign,
+    events::{reaction::Relation as AnnotationRelation, AnyMessageLikeEventContent},
+    OwnedEventId, OwnedUserId, TransactionId, UInt,
+};
+use tracing::{debug, error};
+
+use super::{Joined, Room};
+use crate::{
+    event_handler::EventHandlerDropGuard,
+    room::{self, MessagesOptions},
+    Result,
+};
+
+mod event_handler;
+mod event_item;
+mod virtual_item;
+
+use self::event_handler::{handle_back_paginated_event, handle_live_event, handle_local_event};
+pub use self::{
+    event_item::{
+        EventTimelineItem, Message, PaginationOutcome, ReactionDetails, TimelineDetails,
+        TimelineItemContent, TimelineKey,
+    },
+    virtual_item::VirtualTimelineItem,
+};
+
+/// A high-level view into a regular¹ room's contents.
+///
+/// ¹ This type is meant to be used in the context of rooms without a
+/// `room_type`, that is rooms that are primarily used to exchange text
+/// messages.
+#[derive(Debug)]
+pub struct Timeline {
+    inner: TimelineInner,
+    room: room::Common,
+    start_token: Mutex<Option<String>>,
+    _end_token: Mutex<Option<String>>,
+    _event_handler_guard: EventHandlerDropGuard,
+}
+
+#[derive(Clone, Debug, Default)]
+struct TimelineInner {
+    items: MutableVec<Arc<TimelineItem>>,
+    // Reaction event / txn ID => sender and reaction data
+    reaction_map: Arc<Mutex<HashMap<TimelineKey, (OwnedUserId, AnnotationRelation)>>>,
+}
+
+impl Timeline {
+    pub(super) fn new(room: &room::Common) -> Self {
+        let inner = TimelineInner::default();
+
+        let handle = room.add_event_handler({
+            let inner = inner.clone();
+            move |event, encryption_info: Option<EncryptionInfo>, room: Room| {
+                let inner = inner.clone();
+                async move {
+                    handle_live_event(event, encryption_info, room.own_user_id(), &inner);
+                }
+            }
+        });
+        let _event_handler_guard = room.client.event_handler_drop_guard(handle);
+
+        Timeline {
+            inner,
+            room: room.clone(),
+            start_token: Mutex::new(None),
+            _end_token: Mutex::new(None),
+            _event_handler_guard,
+        }
+    }
+
+    /// Add more events to the start of the timeline.
```

Comment:
I wonder if calling this operation *scrolling* would add more clarity.

Reply:
🤷🏼